snuba/datasets/storage.py

```diff
@@ -0,0 +1,95 @@
+from abc import ABC
+from typing import Optional, Sequence
+
+
+from snuba.datasets.dataset_schemas import StorageSchemas
+from snuba.datasets.table_storage import TableWriter
+from snuba.query.query import Query
+from snuba.query.query_processor import QueryProcessor
+from snuba.request.request_settings import RequestSettings
+
+
+class Storage(ABC):
+    """
+    Storage is an abstraction on anything we can run a query onto in our
+    database. This means that it generally represents a Clickhouse table
+    or a view.
+    It provides:
+    - what we need to build the query (the schemas)
+    - it can provide a table writer if we can write on this storage
+    - a sequence of query processors that are applied to the query after
+      the storage is selected.
+
+    There are one or multiple storages per dataset (in the future, there
+    will be multiple per entity). During the query processing a storage
+    is selected and the query focuses on that storage from that point.
+    """
+
+    # TODO: Break StorageSchemas apart. It contains a distinction between write schema and
```

Comment:
StorageSchema provides: read_schema, write_schema, intermediate_schema. This was from when a dataset did not have storages, only schemas.
I think this should just return a Sequence of schemas for DDL operations (like for matviews where there are three schemas) and a dedicated schema for queries and remove StorageSchemas.
The TableWriter already knows which one is the schema to write on. Though I need this step before we can break StorageSchemas otherwise DDL operations do not know what to create and in which order and fitting that change in this PR would be too big.